### PR TITLE
network: Fix IPv6 PreferredSource routes

### PR DESCRIPTION
**Pull request by daxtens**
If an IPv6 route is added with a source address that is still
tentative, the kernel will refuse to install it.

Previously, once we sent the messages to the kernel to add the
addresses, we would immediately proceed to add the routes. The
addresses would usually still be tentative at this point, so
adding static IPv6 routes was broken - see issue #5882.

Now, only begin to configure routes once the addresses are ready,
by restructuring the state machine, and tracking when addresses are
ready, not just added.

Fixes: #5882

### CODE REVIEW

**Comment by yuwata** [comment]
Please provide a testcase for the issue in test/test-network/.

**Comment by yuwata** [comment]
cc @ssahani

**Comment by ssahani** [comment]
overall looks ok Please provide the test case so that it helps us track the issue.

**Comment by daxtens** [comment]
Hi,

Thanks for the feedback - I have added a new commit with a test case.

Regards,
Daniel

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -900,6 +843,68 @@ static int link_request_set_routes(Link *link) {
...
+
+        if (!link->neighbors_configured)
+                return;
+
+        SET_FOREACH(a, link->addresses, i)
+                if (!address_is_ready(a))
+                        return;
+
+        if (!link->addresses_ready) {
+                link->addresses_ready = true;
```
Comment (yuwata) [inline src/network/networkd-link.c]: This flag will never be reset. Please reset this flag at the beginning of `link_request_set_addresses()`.

Comment (daxtens) [reply src/network/networkd-link.c]: Thanks - done with https://github.com/systemd/systemd/pull/11375/commits/4e3d84a865a93796cd41b3a63681b5660788f251#diff-12a3ecdf5af7aa9f8583c27ac442ca43R1092

**Comment by yuwata** [comment]
I confirm that this actually fixes the provided test case, but unfortunately, breaks other route or routing policy related tests.
```
======================================================================
FAIL: test_ip_route (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 688, in test_ip_route
    self.assertRegex(output, 'static')
AssertionError: Regex didn't match: 'static' not found in '192.168.0.0/24 proto kernel scope link src 192.168.0.15'

======================================================================
FAIL: test_ip_route_blackhole_unreachable_prohibit (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 710, in test_ip_route_blackhole_unreachable_prohibit
    self.assertRegex(output, 'blackhole')
AssertionError: Regex didn't match: 'blackhole' not found in 'default via 192.168.43.1 dev wlp1s0 proto dhcp src 192.168.43.187 metric 1024 \n192.168.43.0/24 dev wlp1s0 proto kernel scope link src 192.168.43.187 \n192.168.43.1 dev wlp1s0 proto dhcp scope link src 192.168.43.187 metric 1024'

======================================================================
FAIL: test_ip_route_gateway (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 737, in test_ip_route_gateway
    self.assertRegex(output, 'default')
AssertionError: Regex didn't match: 'default' not found in ''

======================================================================
FAIL: test_ip_route_gateway_on_link (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 759, in test_ip_route_gateway_on_link
    self.assertRegex(output, 'default')
AssertionError: Regex didn't match: 'default' not found in ''

======================================================================
FAIL: test_ip_route_tcp_window (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 726, in test_ip_route_tcp_window
    self.assertRegex(output, 'initcwnd 20')
AssertionError: Regex didn't match: 'initcwnd 20' not found in 'default via 192.168.43.1 dev wlp1s0 proto dhcp src 192.168.43.187 metric 1024 \n192.168.43.0/24 dev wlp1s0 proto kernel scope link src 192.168.43.187 \n192.168.43.1 dev wlp1s0 proto dhcp scope link src 192.168.43.187 metric 1024'

======================================================================
FAIL: test_routing_policy_rule (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 612, in test_routing_policy_rule
    self.assertRegex(output, '111')
AssertionError: Regex didn't match: '111' not found in '0:\tfrom all lookup local \n1000:\tfrom all lookup [l3mdev-table] \n32766:\tfrom all lookup main \n32767:\tfrom all lookup default'

======================================================================
FAIL: test_routing_policy_rule_invert (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 646, in test_routing_policy_rule_invert
    self.assertRegex(output, '111')
AssertionError: Regex didn't match: '111' not found in '0:\tfrom all lookup local \n1000:\tfrom all lookup [l3mdev-table] \n32766:\tfrom all lookup main \n32767:\tfrom all lookup default'

======================================================================
FAIL: test_routing_policy_rule_port_range (__main__.NetworkdNetWorkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./systemd-networkd-tests.py", line 629, in test_routing_policy_rule_port_range
    self.assertRegex(output, '111')
AssertionError: Regex didn't match: '111' not found in '0:\tfrom all lookup local \n1000:\tfrom all lookup [l3mdev-table] \n32766:\tfrom all lookup main \n32767:\tfrom all lookup default'

----------------------------------------------------------------------
Ran 67 tests in 815.683s

FAILED (failures=8, expected failures=1)
```

**Comment by daxtens** [comment]
OK, I will look in to that. I did see some failures in the test suite, but I thought they were localised to my machine because some of the tests were failing before the commit as well, and the CI tests passed. I will investigate further and revise the patches.

**Comment by yuwata** [comment]
If I understand correctly, when RA is enabled but no RA exists, then it takes `NDISC_TIMEOUT_NO_RA_USEC = NDISC_ROUTER_SOLICITATION_INTERVAL * NDISC_MAX_ROUTER_SOLICITATIONS = 12sec` to set `ndisc_configured` flag. But, in the test script, `networkd` only runs ~5sec. So, within the duration, `link_check_ready()` does not start to set route or routing policy.

Is it necessary to wait for RA to be configured?

**Comment by yuwata** [comment]
Before this PR, even if `IPv6AcceptRA=no` is not specified, route or routing policy are set soon. Of course, the link does not become 'configured' state in `networkctl` at that time. But, I guess almost all functionalities are actually configured. But this blocks 12 seconds...

**Comment by daxtens** [comment]
Yes, I think your analysis is right. I think it might be possible to reorder the tests in link_check_ready() so as to get the routes installed earlier; I'll have a go at that and let you know how I go.

**Comment by daxtens** [comment]
So I've re-ordered link_check_ready() so that the slow link-local addressing checks are done towards the end. As far as I can see, nothing that is configured earlier can sensibly depend on the link local addresses, and the tests seem to pass within a reasonable time period now.

**Pull request merged**